crates/polars-mem-engine/src/executors/scan/ipc.rs

```diff
@@ -65,211 +65,109 @@ impl IpcExec {
-                        // row limit, passing the total row limit to each
-                        // individual reader helps.
-                        n_rows.map(|n| {
-                            n.saturating_sub(already_read_in_sequence)
-                                .try_into()
-                                .unwrap()
-                        }),
-                    )
-                    .with_row_index(self.file_options.row_index.clone())
-                    .with_projection(projection.clone())
-                    .memory_mapped(memory_mapped)
-                    .finish()?;
+        let dfs = if let Some(predicate) = self.predicate.clone() {
+            let predicate = phys_expr_to_io_expr(predicate);
+            let predicate = Some(predicate.as_ref());
 
-                row_counter
-                    .write()
-                    .unwrap()
-                    .write(index, df.height().try_into().unwrap());
+            POOL.install(|| {
+                dfs.into_par_iter()
+                    .map(|mut df| {
+                        apply_predicate(&mut df, predicate, true)?;
+                        Ok(df)
+                    })
+                    .collect::<PolarsResult<Vec<_>>>()
+            })?
+        } else {
+            dfs
+        };
 
-                Ok((index, df))
-            })
-            .collect::<PolarsResult<Vec<_>>>()?;
+        accumulate_dataframes_vertical(dfs)
+    }
 
-        finish_index_and_dfs(
-            index_and_dfs,
-            row_counter.into_inner().unwrap(),
-            self.file_options.row_index.as_ref(),
-            row_limit,
-            self.predicate.as_ref(),
-        )
+    fn read_sync(&mut self) -> PolarsResult<DataFrame> {
+        let paths = self.paths.clone();
+        self.read_impl(move |i| std::fs::File::open(&paths[i]).map_err(Into::into))
     }
 
     #[cfg(feature = "cloud")]
-    async fn read_async(&mut self, verbose: bool) -> PolarsResult<DataFrame> {
-        use futures::stream::{self, StreamExt};
-        use futures::TryStreamExt;
-
-        /// See https://users.rust-lang.org/t/implementation-of-fnonce-is-not-general-enough-with-async-block/83427/3.
-        trait AssertSend {
-            fn assert_send<R>(self) -> impl Send + stream::Stream<Item = R>
-            where
-                Self: Send + stream::Stream<Item = R> + Sized,
-            {
-                self
-            }
-        }
-
-        impl<T: Send + stream::Stream + Sized> AssertSend for T {}
-
-        let n_rows = self
-            .file_options
-            .n_rows
-            .map(|limit| limit.try_into().unwrap());
-
-        let row_limit = n_rows.unwrap_or(IdxSize::MAX);
-
-        let row_counter = RwLock::new(ConsecutiveCountState::new(self.paths.len()));
+    async fn read_async(&mut self) -> PolarsResult<DataFrame> {
+        use polars_io::file_cache::init_entries_from_uri_list;
```

Comment:
Changed `read_async` to dispatch to the sync impl - this works because we are currently using the file download cache for async IPC.

We don't have it right now but eventually I also want a better async IPC implementation that only downloads the parts of the file it needs.